### ISSUE TITLE
Getting SIGSEGV with or tools v7.3 on mac os mojave 10.14

### ISSUE THREAD

**Issue by pardeep632** [google/or-tools#1831]
Crash Logs 

A fatal error has been detected by the Java Runtime Environment:
SIGSEGV (0xb) at pc=0x0000000124b88cb7, pid=921, tid=0x0000000000005103
JRE version: Java(TM) SE Runtime Environment (8.0_172-b11) (build 1.8.0_172-b11)
Java VM: Java HotSpot(TM) 64-Bit Server VM (25.172-b11 mixed mode bsd-amd64 compressed oops)
Problematic frame:
C  [libjniortools.jnilib+0x4fcb7]  _ZN7JNIEnv_15DeleteGlobalRefEP8_jobject+0x7
Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
If you would like to submit a bug report, please visit:
 http://bugreport.java.com/bugreport/crash.jsp
The crash happened outside the Java Virtual Machine in native code.
 See problematic frame for where to report the bug.

---------------  T H R E A D  ---------------

Current thread (0x00007fc218847000):  JavaThread "Finalizer" daemon [_thread_in_native, id=20739, stack(0x000070001041a000,0x000070001051a000)]

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x00000000000000c0

Registers:
RAX=0x0000000000000010, RBX=0x00007fc21e424310, RCX=0x0000000000000000, RDX=0x000000000004b740
RSP=0x0000700010519610, RBP=0x0000700010519610, RSI=0x00007fc219c84e48, RDI=0x00007fc2183e39f8
R8 =0x0000000000000000, R9 =0x00007fc21e0b3a90, R10=0x0000000000000001, R11=0x0000000000000001
R12=0x00007fc21d2ca200, R13=0x00007fc21d2ca2f8, R14=0x00007fc21e441720, R15=0x00007fc21e441828
RIP=0x0000000124b88cb7, EFLAGS=0x0000000000010206, ERR=0x0000000000000004
  TRAPNO=0x000000000000000e

Top of Stack: (sp=0x0000700010519610)
0x0000700010519610:   0000700010519620 0000000124ba11d3
0x0000700010519620:   0000700010519640 0000000124ba1385
0x0000700010519630:   0000000124c84bf8 00007fc21e424310
0x0000700010519640:   0000700010519660 0000000124ba133e
0x0000700010519650:   0000000124c84bf8 00007fc21e43b590
0x0000700010519660:   0000700010519680 0000000124be45e2
0x0000700010519670:   0000000124c84bf8 00007fc21e441750
0x0000700010519680:   0000700010519740 000000012b25c4e8
0x0000700010519690:   00007000105196d0 0000000000000000
0x00007000105196a0:   00007fc21d2ca200 00007fc21d2ca2f8
0x00007000105196b0:   00007fc21d2caa18 0000000124c62080
0x00007000105196c0:   0000000000000000 0000000000000000
0x00007000105196d0:   0000000000000000 0000000000000000
0x00007000105196e0:   0000000000000000 0000000124c62080
0x00007000105196f0:   0000000000000000 0000000000000000
0x0000700010519700:   0000000000000000 0000000000000000
0x0000700010519710:   0000000000000000 00007fc21d2ca200
0x0000700010519720:   0000000000000000 0000000124a35f68
0x0000700010519730:   00007000105197e8 00007fc218847000
0x0000700010519740:   0000700010519760 0000000124b81446
0x0000700010519750:   0000000124b81430 0000000124a35f68
0x0000700010519760:   00007000105197c0 00000001113d7667
0x0000700010519770:   00007000105197b0 00000001113d7382
0x0000700010519780:   0000700010519780 0000000124a35f68
0x0000700010519790:   00007000105197e8 0000000124a46270
0x00007000105197a0:   0000000000000000 0000000124a35f68
0x00007000105197b0:   0000000000000000 00007000105197e0
0x00007000105197c0:   0000700010519840 00000001113c72bd
0x00007000105197d0:   00000007ad201e58 00000001113d0d18
0x00007000105197e0:   00007fc21d2ca200 00000007ade603b0
0x00007000105197f0:   00000017c2e70609 00000007ade603b0
0x0000700010519800:   00007000105197f0 0000000124948fd9 

Instructions: (pc=0x0000000124b88cb7)
0x0000000124b88c97:   48 8b 80 18 07 00 00 5d ff e0 66 2e 0f 1f 84 00
0x0000000124b88ca7:   00 00 00 00 0f 1f 44 00 00 55 48 89 e5 48 8b 07
0x0000000124b88cb7:   48 8b 80 b0 00 00 00 5d ff e0 66 2e 0f 1f 84 00
0x0000000124b88cc7:   00 00 00 00 0f 1f 44 00 00 55 48 89 e5 48 8b 07 

Register to memory mapping:

RAX=0x0000000000000010 is an unknown value
RBX=0x00007fc21e424310 is an unknown value
RCX=0x0000000000000000 is an unknown value
RDX=0x000000000004b740 is an unknown value
RSP=0x0000700010519610 is pointing into the stack for thread: 0x00007fc218847000
RBP=0x0000700010519610 is pointing into the stack for thread: 0x00007fc218847000
RSI=0x00007fc219c84e48 is a global jni handle
RDI=0x00007fc2183e39f8 is an unknown value
R8 =0x0000000000000000 is an unknown value
R9 =0x00007fc21e0b3a90 is an unknown value
R10=0x0000000000000001 is an unknown value
R11=0x0000000000000001 is an unknown value
R12=0x00007fc21d2ca200 is an unknown value
R13=0x00007fc21d2ca2f8 is an unknown value
R14=0x00007fc21e441720 is an unknown value
R15=0x00007fc21e441828 is an unknown value

Stack: [0x000070001041a000,0x000070001051a000],  sp=0x0000700010519610,  free space=1021k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libjniortools.jnilib+0x4fcb7]  _ZN7JNIEnv_15DeleteGlobalRefEP8_jobject+0x7
C  [libjniortools.jnilib+0x681d3]  _ZN14GlobalRefGuardD2Ev+0x13
C  [libjniortools.jnilib+0x68385]  _ZNSt3__114__shared_count16__release_sharedEv+0x25
C  [libjniortools.jnilib+0x6833e]  _ZNSt3__119__shared_weak_count16__release_sharedEv+0xe
C  [libjniortools.jnilib+0xab5e2]  _ZNSt3__110__function6__funcIZ86Java_com_google_ortools_constraintsolver_mainJNI_RoutingModel_1registerTransitCallbackE4$_91NS_9allocatorIS2_EEFxxxEE18destroy_deallocateEv+0x12
C  [libortools.dylib+0x4de4e8]  _ZN19operations_research12RoutingModelD2Ev+0x4f8
C  [libjniortools.jnilib+0x48446]  Java_com_google_ortools_constraintsolver_mainJNI_delete_1RoutingModel+0x16
j  com.google.ortools.constraintsolver.mainJNI.delete_RoutingModel(J)V+0
j  com.google.ortools.constraintsolver.RoutingModel.delete()V+25
j  com.google.ortools.constraintsolver.RoutingModel.finalize()V+1
J 3634 C2 java.lang.ref.Finalizer.access$100(Ljava/lang/ref/Finalizer;Lsun/misc/JavaLangAccess;)V (6 bytes) @ 0x0000000111c9ef10 [0x0000000111c9ed20+0x1f0]
j  java.lang.ref.Finalizer$FinalizerThread.run()V+45
v  ~StubRoutines::call_stub
V  [libjvm.dylib+0x2f0b3a]
V  [libjvm.dylib+0x2f12de]
V  [libjvm.dylib+0x2f148a]
V  [libjvm.dylib+0x34bdb1]
V  [libjvm.dylib+0x570a2f]
V  [libjvm.dylib+0x57212a]
V  [libjvm.dylib+0x48f2f6]
C  [libsystem_pthread.dylib+0x333d]  _pthread_body+0x7e
C  [libsystem_pthread.dylib+0x62a7]  _pthread_start+0x46
C  [libsystem_pthread.dylib+0x2425]  thread_start+0xd
C  0x0000000000000000

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  com.google.ortools.constraintsolver.mainJNI.delete_RoutingModel(J)V+0
j  com.google.ortools.constraintsolver.RoutingModel.delete()V+25
j  com.google.ortools.constraintsolver.RoutingModel.finalize()V+1
J 3634 C2 java.lang.ref.Finalizer.access$100(Ljava/lang/ref/Finalizer;Lsun/misc/JavaLangAccess;)V (6 bytes) @ 0x0000000111c9ef10 [0x0000000111c9ed20+0x1f0]
j  java.lang.ref.Finalizer$FinalizerThread.run()V+45
v  ~StubRoutines::call_stub

Code Snippet 

```
private fun orToolsSolver(dataModel: DataModel): HashMap<Int, ArrayList<Long>> {
        val resourceConstraints: ResourceConstraints = get()
        val dimensionConstraints: DimensionConstraints = get()
        val penaltiesConstraints: PenaltiesAndDroppingVisitConstraints = get()

        val dd = dataModel.timeDistancePair.first

        val manager = RoutingIndexManager(dd.size, dataModel.mVehicleList.size, dataModel.depot)
        val routing = RoutingModel(manager)

        val transitCallbackIndex = routing.registerTransitCallback { fromIndex: Long, toIndex: Long ->
            // Convert from routing variable Index to user NodeIndex.
            val fromNode = manager.indexToNode(fromIndex)
            val toNode = manager.indexToNode(toIndex)
            dd[fromNode][toNode]
        }
        routing.setArcCostEvaluatorOfAllVehicles(transitCallbackIndex)

        val demandCallbackIndex = routing.registerUnaryTransitCallback { fromIndex: Long ->
            // Convert from routing variable Index to user NodeIndex.
            val fromNode = manager.indexToNode(fromIndex)
            dataModel.demands[fromNode]
        }
        routing.addDimensionWithVehicleCapacity(
            demandCallbackIndex, 0, // null capacity slack
            dataModel.vehicleCapacities.toLongArray(), // vehicle maximum capacities
            true, // start cumul to zero
            "Capacity"
        )

        resourceConstraints.applyResourceConstraint(dataModel, manager, routing)
        dimensionConstraints.applyDimensionConstraint(dataModel, manager, routing)
        penaltiesConstraints.applyPenaltyConstraint(dataModel, manager, routing)

        val searchParameters = main.defaultRoutingSearchParameters()
            .toBuilder()
            .setFirstSolutionStrategy(FirstSolutionStrategy.Value.PATH_CHEAPEST_ARC)
            .build()

        val solution = routing.solveWithParameters(searchParameters)

        val ret = returnResult(routing, manager, solution, dataModel.mVehicleList.size, dataModel.demands)
        return ret
    }

// ResouceConstraints File

private fun applyResourceConstraint(dataModel: DataModel, manager: RoutingIndexManager,
                                       routing: RoutingModel) {
        val map = ConstraintsUtility.prepareCustomerToVehiclesMap(dataModel, masterDataDao)
        dataModel.nodeToAllowedDriversMap = map
        for (index in 0 until dataModel.customerMap.size) {
            // here +1 to bypass depot i.e. there is no restriction on depot
            val pickupIndex = manager.nodeToIndex(index + 1)
            routing.setAllowedVehiclesForIndex(dataModel.nodeToAllowedDriversMap[index], pickupIndex)
        }
    }


// Dimension Constraint File

private fun applyDimensionConstraint(
        dataModel: DataModel,
        manager: RoutingIndexManager,
        routing: RoutingModel
    ) {
        val maxCeilingAmount = dataModel.ceilingAmount.max() ?: 0
        val vehicleNumber = dataModel.mVehicleList.size
        for (i in 0 until vehicleNumber) {
            val earnedAmount = maxCeilingAmount - dataModel.ceilingAmount[i]
            routing.setFixedCostOfVehicle(earnedAmount, i)
        }
    }


// Penalty Constraints File

private fun addPenaltyConstraints(dataModel: DataModel, manager: RoutingIndexManager, routing: RoutingModel) {
        val distMat = dataModel.timeDistancePair.first
        val noOfNodes = distMat.size // including depot
        var maxNodeDistance = 0L
        distMat.forEach {
            val max = it.max() ?: 0
            maxNodeDistance = if (maxNodeDistance < max) max else maxNodeDistance
        }

        val basePenalty = getbasePenality(dataModel, maxNodeDistance, noOfNodes)

        for (i in 1 until noOfNodes) {
            val penalty = basePenalty + dataModel.nodePriority[i] * maxNodeDistance
            dataModel.penalty.add(penalty)
            routing.addDisjunction(longArrayOf(manager.nodeToIndex(i)), penalty)
        }
    }
```

**Comment by Mizux** [comment]
I would say the lambda function for the registerTransitCallback is garbage collected too early (i.e. ref count to zero). 
could you try to associate the lambda to a val (i.e. making it a "lvalue reference")

**Comment by pardeep632** [comment]
I changed it like this 

```
val longBinaryOperator = LongBinaryOperator { left, right ->
            val fromNode = manager!!.indexToNode(left)
            val toNode = manager!!.indexToNode(right)
            dd[fromNode][toNode]
        }
        val transitCallbackIndex = routing.registerTransitCallback(longBinaryOperator)
```

But still giving same issue

**Comment by nyilmaz** [comment]
This is issue also on linux.

**Comment by nyilmaz** [comment]
It seems manually calling delete on both `RoutingIndexManager` and `RoutingModel` may have solved this issue. By doing this I think we prevent GC to delete dangling `RoutingModel` objects.

**Comment by Mizux** [comment]
I think it is related to #2466 which have been fixed on master (should be available in the incoming v9.1)
